src/types.rs

```diff
@@ -1,51 +1,154 @@
 //! Concrete node and edge types used in the registry.
 
+extern crate num_traits;
+extern crate quickcheck;
+
+use num_traits::Zero;
+use quickcheck::{Arbitrary, Gen};
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::ops::Add;
+
 /// The type of a node.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub enum NodeType {
     /// A user, eg. contributor, project member etc.
     User,
     /// A project with users as members and contributors.
     Project,
 }
 
+impl Arbitrary for NodeType {
+    fn arbitrary<G: Gen>(g: &mut G) -> Self {
+        if g.next_u32() % 2 == 0 {
+            Self::User
+        } else {
+            Self::Project
+        }
+    }
+}
+
 /// Node data.
-type NodeData = NodeType;
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct NodeData<W> {
+    /// The type for this node.
+    pub node_type: NodeType,
+    /// The total contributions by this user, to *all* projects, if any.
+    pub total_contributions: Option<u32>,
+    pub rank: NodeRank<W>,
+}
+
+impl<W> Arbitrary for NodeData<W>
+where
+    W: Arbitrary,
+{
+    fn arbitrary<G: Gen>(g: &mut G) -> Self {
+        NodeData {
+            node_type: Arbitrary::arbitrary(g),
+            total_contributions: Arbitrary::arbitrary(g),
+            rank: Arbitrary::arbitrary(g),
+        }
+    }
+}
 
 /// The type of an edge.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
 pub enum EdgeType {
-    /// Contribution from a user to a project. This edge type is
-    /// used bi-directionally, between project and user.
-    Contribution,
-    /// Membership of a user in a project, project <-> user.
-    ProjectMembership,
+    /// Contribution from a project to a user.
+    Contrib,
+    /// Contribution from a user to a project.
+    ContribStar,
```

Comment:
Can we call this something meaningful?

Reply:
I have tried to use a naming scheme as close as possible to the way the paper also describes the edge types. In the paper we have this unicode "circle" symbol, like in `contrib⚬`, which I didn't know how to vocalise in plain english. Usually when you want to express the dual for something you tend to also use "star" (as in `contrib*`) or "prime" (as in `contrib'`). I am personally neutral -- I have opted for the former, but I am open to suggestions.

Reply:
It's `co-contrib` :relaxed: 

But I think I'd prefer something readable for someone who hasn't read the paper in detail, so maybe the names I had in https://github.com/oscoin/graph-api/pull/6#pullrequestreview-287939959 ?

Reply:
```
    /// Contribution, user -> project, with strength metric.
    ContributionToProject(u32),
    /// Contribution, project -> user, with strength metric.
    ContributionFromUser(u32),
```

Reply:
Yup, not a problem -- as long as there exist a 1:1 correspondence with the edge types in the paper, I am happy to use "developer-friendly" types ;)

Reply:
```
  /// Contribution, user -> project, with strength metric.
    ContributionToProject(u32),
    /// Contribution, project -> user, with strength metric.
    ContributionFromUser(u32),
```

That's not gonna work my friend, as if you embed data into this enum, you cannot use this as a key in the `edge_weights` `HashMap` 😉

Reply:
Yeah that's fine - either way you need two sets of data: one for the tag and one for the content (strength)

Reply:
What I mean by "either way" is you could keep that enum, and have another one with the tag only:

```rust
enum Foo { ContribTo(u32), ContribFrom(u32) }
impl Foo { fn tag(&self) -> FooTag { ... } }
enum FooTag { ContribTo, ContribFrom }
```